plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java

```diff
@@ -410,6 +424,33 @@ private static <K, V> V getWithValueHolder(Cache<K, AtomicReference<V>> cache, K
         return value;
     }
 
+    private static <K, V> V getIncrementally(
+            Cache<K, AtomicReference<V>> cache,
+            K key,
+            Predicate<V> isSufficient,
+            Supplier<V> loader,
+            Function<V, V> incrementalLoader,
+            BinaryOperator<V> merger)
+    {
+        AtomicReference<V> valueHolder = uncheckedCacheGet(cache, key, AtomicReference::new);
+        V oldValue = valueHolder.get();
+        if (oldValue != null && isSufficient.test(oldValue)) {
+            return oldValue;
+        }
+
+        V newValue = oldValue == null ? loader.get() : incrementalLoader.apply(oldValue);
+
+        verifyNotNull(newValue, "loader returned null for %s", key);
+
+        V merged = merger.apply(oldValue, newValue);
+        if (!valueHolder.compareAndSet(oldValue, merged)) {
+            // if the value changed in the valueHolder, we only add newly loaded value to be sure we have up-to-date value
+            valueHolder.accumulateAndGet(newValue, merger);
+        }
+        // return merged value in any case so that we don't depend on the 'always-growing' property of the valueHolder
+        return merged;
```

Comment:
I know what "always-growing" means, but I am not convinced future readers of the code would.
I'd remove the comment.